### ISSUE TITLE
Investigate push notifications with Pushy

### ISSUE THREAD

**Issue by swentel** [swentel/indigenous-android#214]
Not sure why I deleted the original issue.

@EdwardHinkle has ideas around this: see https://github.com/EdwardHinkle/indigenous-ios/issues/15 and https://github.com/EdwardHinkle/indigenous-ios/issues/227

However, I'm not 100% sure about that setup since it relies on polling and the server storing the access token too. I'd rather have it 'instant', e.g. when new content arrives, especially in my microsub notifications channel.

- edit - it seems the second issue is more about polling, the first one is more or less what I came with myself :)

Current brainstorm in my mind, using Pushy.me - https://pushy.me/
(Pushy doesn't rely on GCM which is great as I'm on LineageOS now)

-  setup a site where people register using IndieAuth with their site. 
- When a device wants to register, it pings the site first to see if their site exists. 
- If so, send the Pushy token to the site.
- On the site, show the registered device along with a token. This is not the Pushy token
- This token can be used to ping the site with a request containing a payload
- When this request comes in, the site pushes it, using the Pushy Rest API, and Pushy sends it to the registered devices

For custom servers:

- allow adding a custom uri, and optionally username and password
- subscribe to something - need to figure out what exactly this is

**Comment by EdwardHinkle** [comment]
Very cool! It does sound similar to https://github.com/EdwardHinkle/indigenous-ios/issues/15 essentially allowing someone to send themselves push notifications about anything.

As you mentioned https://github.com/EdwardHinkle/indigenous-ios/issues/227 is really something different. That is about saying "I want a notification anytime this channel gets new content". It would start by polling but my eventual hope is that a Microsub server would support WebSub notifications for channel content so that wouldn't always have to be a pull but WebSub could make it a push.

But besides the pull/push I think the key question is where the content originates: is it generic content being pushed in or is it just content when a Microsub channel is updated. I am interested in both, hence my separate tickets 😃

**Comment by swentel** [comment]
Hmm right. Anyone could send anything (well, limited to a few properties I guess), on any given moment. In a way, that might be dangerous. However, I'm sure a) Pushy has a throttling service and b) I'd built that in the service as well :)

> I want a notification anytime this channel gets new content

Right, so the notification would be a payload saying something like 'Hey, you have 3 new items in your {insert channel name} channel!'

In a way, that's more or less what I would push out myself:

- 'x has replied on y' - basically things from the notifications channel, being individual payloads
- 'There are x new items in channel y'

Since I control my own microsub server, I know when something comes into my notifications channel, or when a feed has new content. And since I control my own server, I could send out random push notifications too e.g. immediately when a webmention comes in. It's just at a different point in time then instead of waiting for the content to come into my notifications channel (although, that's relatively instant heh).

Now users on aperture.p3k.io don't have that, so I see where this polling might come in handy for them. Unless we provide a patch for Aperture to have a setting to enter their IndiePush keys, and aperture then sends a ping when new content came in.

When I think of it some more, AFAICS, for both issues you have open, the server on which someone registers is actually the same. It just has a built-in service that does the polling and sends you a 'hey, you have x new items in channel y'. However, two things that come to mind:

- I don't see how WebSub comes in play at some point, then the microsub server becomes responsible for the push no (or the ping to the server which then pushes) ?
- I'd rather not store the access token with which you authenticated in Indigenous since the scope of that token would simply be too big I think (I currently send all scopes in Indigenous of both micropub and microsub, so way to many privileges), which is why I'd rather have you authenticate on this server/service with only a 'read' scope for the polling part.

Need to sleep, lot's to think about. I was more or less against the polling thing, but I can see where it's useful, especially for users using hosted versions of aperture (or other microsub server) which they don't control. 

I also don't want us to start writing different services, that would be a total waste of time and efforts, so for now, what I can see me building is:

- a service/site where you register with IndieAuth
- enable auto polling which sends a 'standard' message after you get a 'read' access token from your microsub server
- allow to send a payload to the server with an API token from this server, with limited properties

Then, both iOS and Android can send the device token with the URL (the URL must be an exact match with which you registered, or it will not register) to this service, and then all the magic can start :)

This should cover both your cases then no ?

**Comment by EdwardHinkle** [comment]
So a couple of things. Indigenous for iOS is going to go through some fairly big changes this year. One part is a web/server component that will definitely be handling some of this stuff. Another thing is, the name “Indigenous for iOS” will probably be going away, and you will be free to just refer to Indigenous for Android as “Indigenous”. 🙂

There will be a variety of things we’ll want to discuss/handle as far as website, etc. for the most part, we could just leave it as is (minus the iOS page) and I can update the server whenever you need, or whenever you wish, you could also set up your own site with the same GitHub code (or different, up to you).

I don’t want to go too much into it because I’m not really ready to talk publicly about it (despite posting it here in this issue publicly lol). 

TL;DR: You’ll be free to do whatever you want with the Indigenous branding, I’m happy to do anything on my side to continue supporting you as much as you still want me to Server/Code-wise. Ultimately, if possible I’m happy to let Indigenous (“for Android”) traffic run through the new server conceptually, but obviously the server will be built with specific changes in mind that may or may not work with how you’d like to do this. So definitely don’t hold back on my account, if you get this set up for Indigenous for Android, awesome! If you haven’t gotten it set up when I transition Indigenous for iOS and introduce the server/notifications component, I’m happy to talk things through and see if behind the scenes it would work for you to piggyback off of the existing infrastructure or not.

**Comment by swentel** [comment]
Oh good to know in a way, curious about the changes :)

I guess it probably makes more sense to 'split' things then at some point where I maintain a site myself on a new domain. This site then handles the callback for authentication in the app, but also is used to register devices for push notifications, and who knows, maybe other services as well.

Doing that lets us basically move things forward without 'blocking' or being dependant eachother, works for me :)

**Comment by swentel** [comment]
This is ready for testing :)

I have a new site at https://indigenous.realize.be - the app will use this site also for the auth callback. But it would be great to keep the callback that exists on indigenous.adobe.pub for a couple of months.
I will ping you when there's a new release out, then we can add redirects from indigenous.adobe.pub to indigenous.realize.be.

**Comment by EdwardHinkle** [comment]
@swentel wow, you are quick! Haha yeah, we’ll definitely have the indigenous.abode.pub address for a couple of months as it’ll probably take that long for the transitions I’m working on.

Congrats on the quick progress on this!
